Attempt to update config broker demo for sealed capabilities.

rmn30 opened the pull request:
This is a pretty mechanical attempt at an update by analogy with https://github.com/CHERIoT-Platform/network-stack/commit/95575b74b377d7b6b6d6aecab6b528c6c894662e .  It is probably suboptimal: I suspect ConfigToken should really be three types.  Also the macros in compartment-macros.h need updating to better handle the situation where the declared type of the static sealed thing differs from the actual type due to e.g. a trailing array of variable length.

rmn30 committed:
Attempt to update config broker demo for sealed capabilities.

This is a pretty mechanical attempt at an update by analogy with
https://github.com/CHERIoT-Platform/network-stack/commit/95575b74b377d7b6b6d6aecab6b528c6c894662e
.  It is probably suboptimal: I suspect ConfigToken should really be
three types.  Also the macros in compartment-macros.h need updating to
better handle the situation where the declared type of the static
sealed thing differs from the actual type due to e.g. a trailing array
of variable length.

rmn30 committed:
Fix some typos in configuration broker README.

rmn30 committed:
Split ConfigCapability into different types and use DECLARE_AND_DEFINE_STATIC_SEALED_VALUE_EXPLICIT_TYPE.

Read and Write config capabilities only need a name and are both aliases for ConfigName but sealed with different keys.
The existing ConfigToken type serves for paresr config capabilities.
The new EXPLICIT_TYPE macro removes the need for casts.

rmn30 committed:
Increase sleep in mqtt_stub.

It was previously too short as needs to be long enough to for updates to RGB Led two iterations apart to succeed with a minimum update interval of 1800ms. Presumably was previously working only by luck.

PhilDay-CT commented on configuration_broker/common/config_broker/config_broker.h:
```diff
@@ -48,17 +53,14 @@ struct ConfigToken
-		  uint32_t   notUsed2;                                                 \
 		  const char Name[sizeof(name)];                                       \
 	  },                                                                       \
+	  struct ConfigName,                                                        \
```
To be consistent with the previous declaration could just be "ConfigName" ?

davidchisnall replied:
Normally, we use .h for headers that work in C and C++ modes, .hh for C++-only ones. I’d expect this to be a .hh or say struct.

PhilDay-CT replied:
KK - so line 42 should also say struct then ?

rmn30 replied:
I added the missing `struct` and rebased on main.

PhilDay-CT commented on configuration_broker/common/config_broker/config_broker.h:
```diff
@@ -70,12 +72,13 @@ struct ConfigToken
 		  uint32_t   update_interval;                                          \
 		  const char Name[sizeof(name)];                                       \
 	  },                                                                       \
+	  struct ConfigToken,                                                        \
```
just ConfigToken ?

PhilDay-CT commented on configuration_broker/sonata/provider/mqtt.cc:
```diff
@@ -220,9 +220,9 @@ void __cheri_compartment("provider") provider_run()
 		t               = UnlimitedTimeout;
-		SObj mqttHandle = mqtt_connect(&t,
+		auto mqttHandle = mqtt_connect(&t,
 		                               STATIC_SEALED_VALUE(mqttTestMalloc),
```
Should this now be MALLOC_CAPABILITY(mqttTestMalloc) ?

rmn30 replied:
No, because `MALLOC_CAPABILITY` is defined to be the [default allocator capability for malloc](https://github.com/CHERIoT-Platform/cheriot-rtos/blob/ca9845a80efba06649e1bfb971e443e0da028686/sdk/include/stdlib.h#L97) . It's a bit confusing. We don't currently have an equivalent of `READ_CONFIG_CAPABILITY` and the like for allocator capabilities. Those macros were mainly needed to help with the compiler transition to sealed capabilities before we had the `EXPLICIT_TYPE` versions of the static scaled capability macros for variably sized things. They're not strictly necessary now as plain `STATIC_SEALED_VALUE` will work fine, but I quite like the named macros as they document intent better and may be a useful hook at some point. I don't mind whether we keep them or not. Perhaps we should add them for other uses of static sealed capabilities?

PhilDay-CT reviewed (commented):
Looks good, a few minor queries

rmn30 committed:
Config broker: fix two unintended crashes in parsers.

Both occurred when logging failure to parse a message so went unnoticed.
The logger parser was accidentally reading from a write-only capability due to using the wrong pointer.
The rgb_led parser was attempting to print a string that did not include the null terminator because the capability is resized to the result of strlen in order to emulate the real MQTT provider.
The latter is possibly something the debug printer should handle as per https://github.com/CHERIoT-Platform/cheriot-rtos/issues/450 .

PhilDay-CT reviewed (approved):
Looks good - thanks for all your work on this
